packages/sacado/src/new_design/Sacado_Fad_Exp_GeneralFadTestingHelpers.hpp

```diff
@@ -0,0 +1,130 @@
+// License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301
+// USA
+// Questions? Contact David M. Gay (dmgay@sandia.gov) or Eric T. Phipps
+// (etphipp@sandia.gov).
+//
+// ***********************************************************************
+//
+// @HEADER
+
+#ifndef SACADO_FAD_EXP_GENERALFADTESTINGHELPERS_HPP
+
+#ifndef TEUCHOS_TESTING_HELPERS_HPP
+#include "Teuchos_TestingHelpers.hpp"
+#endif
+
+#define SACADO_FAD_EXP_GENERALFADTESTINGHELPERS_HPP
+
+namespace Sacado {
+  namespace Fad {
+  namespace Exp {
+    template <typename Storage>
+    typename GeneralFad<Storage>::scalar_type relErr( const GeneralFad<Storage> &s1, const GeneralFad<Storage> &s2 );
+  }
+  }
+}
+
+namespace Teuchos {
+  using Sacado::Fad::Exp::relErr;
+}
+
+namespace Sacado {
+  namespace Fad {
+  namespace Exp {
+
+    template <typename Scalar1, typename Scalar2>
+    typename Scalar1::scalar_type
+    relErrFadImpl( const Scalar1 &s1, const Scalar2 &s2 )
+    {
+      typename Scalar1::scalar_type maxRelErr = Teuchos::relErr(s1.val(), s2.val());
+      for (int i=0; i<s1.size(); ++i) {
+        typename Scalar1::scalar_type tmpRelErr = Teuchos::relErr(s1.dx(i), s2.dx(i));
+        if (tmpRelErr >= maxRelErr)
+          maxRelErr = tmpRelErr;
+      }
+      return maxRelErr;
+    }
+
+    template <typename Storage>
+    typename GeneralFad<Storage>::scalar_type relErr( const GeneralFad<Storage> &s1, const GeneralFad<Storage> &s2 ) { return relErrFadImpl(s1.derived(),s2.derived()); }
+  }
+  }
+}
+
+namespace Teuchos {
+  template <typename T1, typename T2>
+  struct TestRelErr<T1,T2,
+                    typename std::enable_if<Sacado::IsExpr<T1>::value ||
+                                            Sacado::IsExpr<T2>::value >::type >
+  {
+    typedef typename Sacado::ScalarType<T1>::type scalarType1;
+    typedef typename Sacado::ScalarType<T2>::type scalarType2;
+    typedef typename std::common_type<scalarType1,scalarType2>::type magnitudeType;
+    static bool eval(
+      const std::string &v1_name,
+      const T1 &v1,
+      const std::string &v2_name,
+      const T2 &v2,
+      const std::string &maxRelErr_error_name,
+      const magnitudeType &maxRelErr_error,
+      const std::string &maxRelErr_warning_name,
+      const magnitudeType &maxRelErr_warning,
+      const Ptr<std::ostream> &out
+    )
+    {
+      using std::endl;
+      typedef Teuchos::ScalarTraits<magnitudeType> SMT;
+      const auto rel_err = Teuchos::relErr( v1.derived(), v2.derived() );
```

Comment:
Can you change this to call relErrFadImpl directly so that it will work for both the old and new designs?